File: scm-java/src/main/scheme/primitives/PrimitiveRecordSetB.java
package scheme;

public class PrimitiveRecordSetB extends Primitive {
    public String name() { return "record-set!"; }

    public String info() {
        return "Syntax: (record-set! record index value)\n" +
               "Library: (scm core)\n" +
               "Description: Sets the record slot at the given index to value.\n" +
               "Example:\n" +
               "  (let ((r (make-record 3))) (record-set! r 0 'hello) (record-ref r 0)) => hello";
    }

    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 3, 3);
        Record r = Value.asRecord(arguments[0]);
        int index = IntegerMath.toInt(arguments[1]);
        r.fields[index] = arguments[2];
        return new Values();
    }
}

File: scm-csharp/scheme/primitives/PrimitiveRecordSetB.cs
namespace scheme;

public class PrimitiveRecordSetB : Primitive
{
    public override string Name() => "record-set!";

    public override string Info() =>
        "Syntax: (record-set! record index value)\n" +
        "Library: (scm core)\n" +
        "Description: Sets the record slot at the given index to value.\n" +
        "Example:\n" +
        "  (let ((r (make-record 3))) (record-set! r 0 'hello) (record-ref r 0)) => hello";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 3, 3);
        Record r = Value.AsRecord(arguments[0]);
        int index = IntegerMath.ToInt(arguments[1]);
        r.Fields[index] = arguments[2];
        return new Values();
    }
}
